src/typeChecker/__tests__/typeErrorChecker.test.ts

```diff
@@ -0,0 +1,984 @@
+    `)
+  })
+
+  it('gets return type correct both with and without braces', () => {
+    const context = mockContext(Chapter.SOURCE_1, Variant.TYPED)
+
+    const code = `((a: number, b: number): number => a + b)(1, 2); // no error
+      ((a: number, b: number): string => a + b)(1, 2); // error
+      ((a: string, b: string): number => {
+        return a + b; // error
+      })('1', '2');
+      ((a: string, b: string): string => { 
+        return a + b; // no error
+      })('1', '2');
+    `
+
+    parse(code, context)
+    expect(parseError(context.errors)).toMatchInlineSnapshot(`
+      "Line 2: Type 'number' is not assignable to type 'string'.
+      Line 4: Type 'string' is not assignable to type 'number'."
+    `)
+  })
+})
+
+describe('type aliases', () => {
+  it('TSTypeAliasDeclaration nodes should be removed from program at end of typechecking', () => {
+    const context = mockContext(Chapter.SOURCE_1, Variant.TYPED)
+
+    const code = `type stringOrNumber = string | number;
+      const x = 1;
+    `
+
+    const program = parse(code, context)
+    expect(program).toMatchSnapshot() // Should not contain TSTypeAliasDeclaration node
+  })
+
+  it('should not be used as variables', () => {
+    const context = mockContext(Chapter.SOURCE_1, Variant.TYPED)
+
+    const code = `type x = string | number;
+      x;
+    `
+
+    parse(code, context)
+    expect(parseError(context.errors)).toMatchInlineSnapshot(`"Line 2: Name x not declared."`)
+  })
+
+  it('should throw errors for type mismatch', () => {
+    const context = mockContext(Chapter.SOURCE_1, Variant.TYPED)
+
+    const code = `type stringOrNumber = string | number;
+      const x: stringOrNumber = true;
+    `
+
+    parse(code, context)
+    expect(parseError(context.errors)).toMatchInlineSnapshot(
+      `"Line 2: Type 'boolean' is not assignable to type 'string | number'."`
+    )
+  })
+
+  it('should throw errors for undeclared types', () => {
+    const context = mockContext(Chapter.SOURCE_1, Variant.TYPED)
+
+    const code = `const x: x = 1;`
+
+    parse(code, context)
+    expect(parseError(context.errors)).toMatchInlineSnapshot(`"Line 1: Type 'x' not declared."`)
+  })
+
+  it('should coexist with variables of the same name', () => {
+    const context = mockContext(Chapter.SOURCE_1, Variant.TYPED)
+
+    const code = `type x = string | number;
+      const x: x = 1;
+    `
+
+    parse(code, context)
+    expect(parseError(context.errors)).toMatchInlineSnapshot(`""`)
+  })
+})
+
+describe('typecasting', () => {
+  it('TSAsExpression nodes should be removed from program at end of typechecking', () => {
+    const context = mockContext(Chapter.SOURCE_1, Variant.TYPED)
+
+    const code = `const x1: string | number = 1;
+      const x2: string = x1 as string;
+    `
+
+    const program = parse(code, context)
+    expect(program).toMatchSnapshot() // Should not contain TSAsExpression node
+  })
+
+  it('only supports downcasting', () => {
```

Comment:
My understanding of downcasting is in the context of OOP, where a parent object is typecast to a child object. What does it mean here since we're only dealing with primitives?

I see things like
```ts
const x3: string = x1 as string; // no error
```
which is a narrowing of types.

At the same time, I see things like
```ts
const x8 = x1 as any; // error
```
which is a widening of types.

What exactly does "down" mean here?

Reply:
The only form of typecasting allowed in Source would be to typecast a union type to its subset (e.g. typecasting `string | number` to `number`), this is what "downcasting" refers to. If needed, I can reword this (e.g. `only supports casting to a sub-type`).

Reply:
Should casting to `any` be handled in a separate test case then?